### ISSUE TITLE
CUDNN Support

### ISSUE THREAD

**Issue by dgcrouse** [jcjohnson/torch-rnn#128]
Feature Request - add option gpu_backend='cudnn' to use CUDNN to replace torch-rnn LSTM layers with their CUDNN implementation. Currently testing in private fork, but would appreciate insight about its use as a drop-in replacement.

**Comment by jcjohnson** [comment]
I agree that this is a good feature, and it would be relatively easy to add.

The complication is that cuDNN layers can only be used on GPU. Even if you use a GPU for training, the saved checkpoints should be runnable on a machine without a GPU; that's why the training script casts to float before saving checkpoints. To keep this same behavior with cuDNN, we really need functionality to convert back and forth between cuDNN layers and torch-rnn layers; this is certainly doable but a bit more annoying.

This has been something that I have been meaning to do for a while, but haven't gotten around to it yet.

**Comment by dgcrouse** [comment]
Yeah, that has been something I've been wrestling with myself. What I was envisioning is just enabling it as a gpu-only option and labeling it as such. I don't know how familiar you are with cuDNN, but one thing I am experiencing in my drop-in replacement is while the "batch first" option should be enabled on the cuDNN LSTM layers (the input is batch x time x dim, if I'm not mistaken), that causes an error when inputting to the View that replaces the temporal convolution layer. Only if that option is set to false does it not give an error.

**Comment by dgcrouse** [comment]
Found a fork here: https://github.com/ngimel/torch-rnn that seems to fix my problems. It seems like there is an active issue within cudnn that deals with converting back to a CPU layer, so it seems like you are not the only one to think so.
